src/Package/Impl/ProjectSystem/Commands/SourceFilesCommand.cs

```diff
@@ -21,34 +29,76 @@
-                return new CommandStatusResult(true, commandText, CommandStatus.Enabled | CommandStatus.Supported);
+                return Task.FromResult(new CommandStatusResult(true, commandText, CommandStatus.Enabled | CommandStatus.Supported));
             }
-            return CommandStatusResult.Unhandled;
+            return Task.FromResult(CommandStatusResult.Unhandled);
         }
 
-        public bool TryHandleCommand(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut) {
+        public async Task<bool> TryHandleCommandAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut) {
             if (commandId == RPackageCommandId.icmdSourceSelectedFiles || commandId == RPackageCommandId.icmdSourceSelectedFilesWithEcho) {
-                var rFiles = nodes.GetSelectedNodesPaths().Where(x =>
-                               Path.GetExtension(x).EqualsIgnoreCase(".r") &&
-                               File.Exists(x));
                 bool echo = commandId == RPackageCommandId.icmdSourceSelectedFilesWithEcho;
-                _interactiveWorkflowProvider.GetOrCreate().Operations.SourceFiles(rFiles, echo);
-                return true;
+                IEnumerable<string> rFiles = Enumerable.Empty<string>();
+
+                IFileSystem fs = new FileSystem();
+                var workflow = _interactiveWorkflowProvider.GetOrCreate();
+                try {
+                    var session = workflow.RSession;
+                    if (session.IsRemote) {
+                        var properties = _configuredProject.Services.ExportProvider.GetExportedValue<ProjectProperties>();
+                        
+                        string projectDir = Path.GetDirectoryName(_configuredProject.UnconfiguredProject.FullPath);
+                        string projectName = properties.GetProjectName();
+                        string remotePath = await properties.GetRemoteProjectPathAsync();
+
+                        var files = nodes.GetSelectedNodesPaths().Where(x =>
+                                   Path.GetExtension(x).EqualsIgnoreCase(RContentTypeDefinition.FileExtension) &&
+                                   fs.FileExists(x));
+                        workflow.ActiveWindow.InteractiveWindow.WriteLine(Resources.Info_CompressingFiles);
+                        
+                        string compressedFilePath = fs.CompressFiles(files, projectDir, new Progress<string>((p) => {
+                            workflow.ActiveWindow.InteractiveWindow.WriteLine(string.Format(Resources.Info_LocalFilePath, p));
+                            string dest = p.MakeRelativePath(projectDir).ProjectRelativePathToRemoteProjectPath(remotePath, projectName);
+                            workflow.ActiveWindow.InteractiveWindow.WriteLine(string.Format(Resources.Info_RemoteDestination, dest));
```

Comment:
Or a method in the `IRInteractiveWorkflowOperations`?

---
In reply to: [80755209](https://github.com/Microsoft/RTVS/pull/2405#discussion_r80755209) [](ancestors = 80755209)